### DISCUSSION TITLE
AOT Compiler?

### DISCUSSION THREAD

**Discussion by nsgnkhibdk2cls0f** [objeck/objeck-lang#106] [General]
https://en.wikipedia.org/wiki/Ahead-of-time_compilation

**Comment by ghost** [comment]
I don't think this is needed. The performance of Objeck is good enough. The binaries produced by AOT is platform dependent. They are not portable. I don't think the performance gained by compiling to native binaries worth it. Only performance critical enterprise systems need something like AOT, where a little more performance gained means a lot of money saved.

**Reply by nsgnkhibdk2cls0f** [reply]
This two weeks passed I learned a bit of C# and .NET 7. I'm astonished because of the ability to compile a managed language to native binary directly. Even though the binary compiled by C# has the extension .exe, it's not a native binary like what compiled by MSVC, it's .NET CIL. This AOT feature allowed me to have a truly native binary.

**Reply by nsgnkhibdk2cls0f** [reply]
I don't care about performance at all. But having a native binary is just cool. I know it's difficult to implement. So far I only know C# and Java (both corporate languages) could do this. Java is a bit different, as it's only compile a dynamic library to speed thing up (partial AOT).

**Reply by nsgnkhibdk2cls0f** [reply]
I do not think Objeck must have AOT nor I need AOT in Objeck, but AOT is a cool idea to know about isn't it? Another different approach compared to Objeck's JIT: https://flerka.github.io/personal-blog/2022-06-21-ahead-of-time-compilation/#aot-vs-jit-compilation

**Reply by nsgnkhibdk2cls0f** [reply]
Another point for AOT is on platforms not allowing JIT like iOS as mentioned on the article above. If you want to see Objeck on iOS then it must supports AOT.

**Reply by nsgnkhibdk2cls0f** [reply]
Java has full AOT too, but has to use GraalVM not the normal JVM: https://www.graalvm.org/22.0/reference-manual/native-image/

**Comment by objeck** [comment]
AOT is interesting, like tracing JITs, but it will offer little performance benefit. The AOT would have to be machine code independent to support x64 and ARM64 targets, two vastly different machine formats. For example, the x64 JIT compiler supports indirect addressing, which is not available on ARM64.

**Reply by ghost** [reply]
AOT is always system dependent @objeck @rhollines

**Reply by ghost** [reply]
Yes, AOT is supposed to be platform dependent @objeck From what I read AOT is simply compile to the native platform's binary format. For example, on Windows x64 AOT will produce a Windows x64 binary, on Linux x64 AOT will produce a Linux x64 binary, so does on Linux arm64,... You can't take the AOT compiled binary from one platform to another platform and expect it to work. There is no such a thing as machine code independent AOT. To be machine code independent, we need bytecode. But Objeck always compiled to bytecode anyway!
